{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f3bff336",
   "metadata": {},
   "outputs": [],
   "source": [
    "#%%appyter init\n",
    "from appyter import magic\n",
    "magic.init(lambda _=globals: _())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "94ca6eb6",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Basic Libraries\n",
    "import os\n",
    "import re\n",
    "from time import sleep\n",
    "import pandas as pd\n",
    "# pd.set_option('max_colwidth', None)\n",
    "\n",
    "# Visualization\n",
    "from dash import Dash, dash_table, dcc, html\n",
    "from dash.dependencies import Input, Output\n",
    "import plotly\n",
    "from plotly import tools\n",
    "import plotly.express as px\n",
    "from jupyter_dash import JupyterDash\n",
    "plotly.offline.init_notebook_mode() # To embed plots in the output cell of the notebook\n",
    "\n",
    "import IPython\n",
    "from IPython.display import HTML, display, Markdown, IFrame\n",
    "\n",
    "# Xena\n",
    "import smart_open\n",
    "\n",
    "# External Code\n",
    "from utils import *"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7879a619",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter hide_code_exec\n",
    "{% do SectionField(\n",
    "    name = 'Xena_Section',\n",
    "    title = 'Gene Expression Data from Xena',\n",
    "    subtitle = '',\n",
    "    img = 'xena.png'\n",
    ")%}\n",
    "\n",
    "{% do SectionField(\n",
    "    name = 'Subgroup_Section',\n",
    "    title = 'Select the Two Subgroups to Generate Differential Gene Expression Signature',\n",
    "    subtitle = 'Click to choose your subgroups. Shift-click to select more than one category for a subgroup.',\n",
    "    img = 'analysis.png'\n",
    ")%}\n",
    "\n",
    "{% do SectionField(\n",
    "    name = 'GS_Libraries_Section',\n",
    "    title = 'Gene Set Libraries',\n",
    "    subtitle = 'Click to select a library to run blitzGSEA on.',\n",
    "    img = 'analysis.png'\n",
    ")%}\n",
    "\n",
    "{% do SectionField(\n",
    "    name = 'Visualization_Section',\n",
    "    title = 'Advanced Parameters for PCA/t-SNE',\n",
    "    subtitle = '',\n",
    "    img = 'analysis.png',\n",
    "    collapse = 'hide'\n",
    ")%}\n",
    "\n",
    "{% do SectionField(\n",
    "    name = 'DEG_Section',\n",
    "    title = 'Advanced Parameters for Generating Differential Gene Expression Signature',\n",
    "    subtitle = '',\n",
    "    img = 'analysis.png',\n",
    "    collapse = 'hide'\n",
    ")%}\n",
    "\n",
    "{% do SectionField(\n",
    "    name = 'blitzGSEA_Section',\n",
    "    title = 'Advanced Parameters for blitzGSEA',\n",
    "    subtitle = '',\n",
    "    img = 'analysis.png',\n",
    "    collapse = 'hide'\n",
    ")%}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bec2be10",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter code_exec\n",
    "{% set DE_params = PMSGStringField(\n",
    "    name = 'DE_params', \n",
    "    label = 'Differential expression parameters', \n",
    "    default = '',\n",
    "    section = 'Xena_Section'\n",
    ") %}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2ed9127f",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter code_exec\n",
    "{% set case_group = CategoryMultiChoiceField(\n",
    "    name = 'category_case',\n",
    "    urlhashkey = 'category',\n",
    "    label = 'Subgroup 1',\n",
    "    description = 'Select category to be in the 1st subgroup',\n",
    "    choices = [''],\n",
    "    default = [''],\n",
    "    section = 'Subgroup_Section'\n",
    ")%}\n",
    "\n",
    "{% set control_group = CategoryMultiChoiceField(\n",
    "    name = 'category_control',\n",
    "    urlhashkey = 'category',\n",
    "    label = 'Subgroup 2',\n",
    "    description = 'Select category to be in the 2nd subgroup',\n",
    "    choices = [''],\n",
    "    default = [''],\n",
    "    section = 'Subgroup_Section'\n",
    ")%}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4f34938f",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter code_exec\n",
    "{% set gs_library = ChoiceField(\n",
    "    name = 'gs_library',\n",
    "    label = 'Gene Set Libraries',\n",
    "    description = 'Select a gene set library',\n",
    "    choices = {'MSigDB_Hallmark': 'h.all.v7.5.1.symbols.gmt', 'MSigDB_C1_Positional': 'c1.all.v7.5.1.symbols.gmt', 'MSigDB_C2_Chemical_Genetic_Perturbations': 'c2.cgp.v7.5.1.symbols.gmt', 'MSigDB_C2_Canonical_Pathways' : 'c2.cp.v7.5.1.symbols.gmt', 'MSigDB_C3_microRNA_Targets': 'c3.mir.v7.5.1.symbols.gmt', 'MSigDB_C3_Transcription_Factor_Targets': 'c3.tft.v7.5.1.symbols.gmt', 'MSigDB_C4_Cancer_Gene_Neighborhoods': 'c4.cgn.v7.5.1.symbols.gmt', 'MSigDB_C4_Cancer_Modules': 'c4.cm.v7.5.1.symbols.gmt', 'MSigDB_C5_Gene_Ontology': 'c5.go.v7.5.1.symbols.gmt', 'MSigDB_C5_Human_Phenotype_Ontology': 'c5.hpo.v7.5.1.symbols.gmt', 'MSigDB_C6_Oncogenic_Signature': 'c6.all.v7.5.1.symbols.gmt', 'MSigDB_C7_ImmuneSigDB': 'c7.immunesigdb.v7.5.1.symbols.gmt', 'MSigDB_C7_Vaccine_Reponse': 'c7.vax.v7.5.1.symbols.gmt', 'MSigDB_C8_Cell_Type_Signature': 'c8.all.v7.5.1.symbols.gmt'},\n",
    "    default = 'MSigDB_Hallmark',\n",
    "    section = 'GS_Libraries_Section'\n",
    ")%}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1ab3bb89",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter code_exec\n",
    "{% set interactive_plot = BoolField(\n",
    "    name = 'interactive_plot', \n",
    "    label = 'Interactive plots?', \n",
    "    default = 'true', \n",
    "    description = 'Check if User wants interactive plots', \n",
    "    section = 'Visualization_Section'\n",
    ")%}\n",
    "\n",
    "{% set visualization_method = ChoiceField(\n",
    "    name = 'visualization_method', \n",
    "    label = 'Visualization Methods', \n",
    "    choices = {'PCA': 'PCA', 't-SNE': 't-SNE'},\n",
    "    default = 'PCA', \n",
    "    description = 'Select a visualization method', \n",
    "    section = 'Visualization_Section'\n",
    ")%}\n",
    "\n",
    "{% set nr_genes = IntField(\n",
    "    name = 'nr_genes', \n",
    "    label = 'Genes for Dimension Reduction', \n",
    "    min = 0, \n",
    "    max = 30000, \n",
    "    default = 2500, \n",
    "    description = 'The maximum number of genes for dimension reduction', \n",
    "    section = 'Visualization_Section'\n",
    ")%}\n",
    "\n",
    "{% set filter_genes = BoolField(\n",
    "    name = 'filter_genes', \n",
    "    label = 'Filter genes?', \n",
    "    default = 'false',\n",
    "    description = 'Check if User wants to filter genes with lowest variances', \n",
    "    section = 'Visualization_Section'\n",
    ")%}\n",
    "\n",
    "{% set low_expression_threshold = FloatField(\n",
    "    name = 'low_expression_threshold', \n",
    "    label = 'Low expression threshold', \n",
    "    default = 0.3, \n",
    "    description = 'Threshold to filter out low expression genes. The value should vary based on the user dataset.', \n",
    "    section = 'Visualization_Section'\n",
    ")%}\n",
    "\n",
    "{% set logCPM_normalization = BoolField(\n",
    "    name = 'logCPM_normalization', \n",
    "    label = 'logCPM normalization?', \n",
    "    default = 'false', \n",
    "    description = 'Check if User wants the dataset to be logCPM-transformed', \n",
    "    section = 'Visualization_Section'\n",
    ")%}\n",
    "\n",
    "{% set log_normalization = LogNormalizationField(\n",
    "    name = 'log_normalization',\n",
    "    label = 'log normalization?',\n",
    "    urlhashkey = 'expression_unit',\n",
    "    default = 'false', \n",
    "    description = 'Check if User wants the dataset to be log-transformed', \n",
    "    section = 'Visualization_Section'\n",
    ")%}\n",
    "\n",
    "{% set z_normalization = BoolField(\n",
    "    name = 'z_normalization', \n",
    "    urlhashkey = 'expression_unit',\n",
    "    label = 'Z normalization?', \n",
    "    default = 'true', \n",
    "    description = 'Check if User wants the dataset to be normalized with Z-normalized method', \n",
    "    section = 'Visualization_Section'\n",
    ")%}\n",
    "\n",
    "{% set q_normalization = BoolField(\n",
    "    name = 'q_normalization', \n",
    "    label = 'Quantile normalization?', \n",
    "    default = 'false', \n",
    "    description = 'Check if User wants the dataset to be normalized with Quantile normalization method', \n",
    "    section = 'Visualization_Section'\n",
    ")%}\n",
    "\n",
    "{% set save_intermediate = BoolField(\n",
    "    name = 'save_intermediate', \n",
    "    label = 'Save intermediate files?', \n",
    "    default = 'false', \n",
    "    description = 'Check if User wants to save intermediate analysis files', \n",
    "    section = 'Visualization_Section'\n",
    ")%}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "965430e8",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter code_exec\n",
    "{% set diff_gex_method = DiffgexChoiceField(\n",
    "    name = 'DiffgexChoiceField',\n",
    "    label = 'Differential expression analysis method',\n",
    "    urlhashkey = 'expression_unit',\n",
    "    choices = {'limma_voom': 'limma_voom', 'characteristic direction': 'characteristic_direction', 'edgeR': 'edgeR', 'DESeq2': 'DESeq2'},\n",
    "    default = 'characteristic direction', \n",
    "    description = 'Set a method to get differentially expressed genes', \n",
    "    section = 'DEG_Section')\n",
    "%}\n",
    "\n",
    "{% set diff_gex_plot_method = ChoiceField(\n",
    "    name = 'diff_gex_plot_method',\n",
    "    label = 'Differential expression analysis plotting method',\n",
    "    choices = {'Volcano plot': 'volcano','MA plot': 'MA_plot'},\n",
    "    default = 'Volcano plot', \n",
    "    description = 'Set a plot method to see differentially expressed genes. Available for limma/edgeR/DESeq2.', \n",
    "    section = 'DEG_Section')\n",
    "%}\n",
    "\n",
    "{% set pvalue_threshold = FloatField(\n",
    "    name = 'pvalue_threshold', \n",
    "    label = 'P-value threshold', \n",
    "    min = 0, \n",
    "    max = 1, \n",
    "    default = 0.05, \n",
    "    step = 0.05,\n",
    "    description = 'Threshold to highlight significantly differentially expressed genes', \n",
    "    section = 'DEG_Section')\n",
    "%}\n",
    "{% set logfc_threshold = FloatField(\n",
    "    name = 'logfc_threshold', \n",
    "    label = 'logFC threshold',\n",
    "    min = 0,\n",
    "    max = 1000,\n",
    "    default = 1.5, \n",
    "    description = 'Threshold to highlight diffentially expressed genes', \n",
    "    section = 'DEG_Section')\n",
    "%}\n",
    "\n",
    "{% set gene_topk = IntField(\n",
    "    name = 'gene_topk', \n",
    "    label = 'Maximum genes for Enrichr', \n",
    "    min = 0, \n",
    "    max = 1000, \n",
    "    default = 500, \n",
    "    description = 'The maximum number of genes discovered by the Characteristic Direction method', \n",
    "    section = 'DEG_Section')\n",
    "%}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d625f9b0",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter code_exec\n",
    "{% set permutation = IntField(\n",
    "    name = 'permutations', \n",
    "    label = 'Permutations', \n",
    "    min = 0, \n",
    "    max = 30000, \n",
    "    default = 2000, \n",
    "    description = 'The number of randomized permutations to estimate ES distributions', \n",
    "    section = 'blitzGSEA_Section')\n",
    "%}\n",
    "\n",
    "{% set min_genes = IntField(\n",
    "    name = 'min_genes', \n",
    "    label = 'Minimum number of genes',\n",
    "    min = 0, \n",
    "    max = 30000, \n",
    "    default = 5, \n",
    "    description = 'The minimum number of genes in geneset', \n",
    "    section = 'blitzGSEA_Section')\n",
    "%}\n",
    "\n",
    "{% set max_genes = IntField(\n",
    "    name = 'max_genes', \n",
    "    label = 'Maximum number of genes',\n",
    "    min = 0, \n",
    "    max = 500, \n",
    "    default = 500, \n",
    "    description = 'The maximum number of genes in geneset', \n",
    "    section = 'blitzGSEA_Section')\n",
    "%}\n",
    "\n",
    "{% set anchors = IntField(\n",
    "    name = 'anchors', \n",
    "    label = 'Anchors',\n",
    "    min = 0, \n",
    "    max = 30000, \n",
    "    default = 20, \n",
    "    description = 'The number of gene set size distributions calculated. Remaining are interpolated.', \n",
    "    section = 'blitzGSEA_Section')\n",
    "%}\n",
    "\n",
    "{% set symmetric = BoolField(\n",
    "    name = 'symmetric', \n",
    "    label = 'Symmetric', \n",
    "    default = 'false', \n",
    "    description = 'Use same distribution parameters for negative and positive ES. If False estimate them separately.', \n",
    "    section = 'blitzGSEA_Section')\n",
    "%}\n",
    "\n",
    "{% set n_table = ChoiceField(\n",
    "    name = 'n_table', \n",
    "    label = 'Number of gene set in full table and top table',\n",
    "    choices = ['10', '11', '12', '13', '14', '15', '16', '17', '18', '19', '20', '21', '22', '23', '24', '25', '26', '27', '28', '29', '30', '31', '32', '33', '34', '35', '36', '37', '38', '39', '40', '41', '42', '43', '44', '45', '46', '47', '48', '49', '50'],\n",
    "    default = '10',\n",
    "    description = 'The top n enriched gene sets displayed in top table', \n",
    "    section = 'blitzGSEA_Section')\n",
    "%}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "85207250",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter code_exec\n",
    "# Parse Xena Browser post data\n",
    "DE_params = r\"\"\"{{DE_params.value}}\"\"\"\n",
    "try:\n",
    "    J = json.loads(DE_params)\n",
    "except: # empty form\n",
    "    display(Markdown(\"<span style=\\\"background-color:#f2dede\\\">Error! No input.</span>\"))\n",
    "    sleep(10)\n",
    "    raise SystemExit(1)\n",
    "expr_dataset_name = J['preferredExpression']['name']\n",
    "xena_host = J['preferredExpression']['host']\n",
    "probemap_name = J['preferredExpression'].get('probemap')\n",
    "unit = J['preferredExpression'].get('unit')\n",
    "expr_data_filename = xenaFileDownloadLink(xena_host, expr_dataset_name)\n",
    "probemap_filename = xenaFileDownloadLink(xena_host, probemap_name)\n",
    "samples = J[\"samples\"]\n",
    "values = J[\"data\"][\"req\"][\"values\"][0]\n",
    "codes = J[\"data\"][\"codes\"]\n",
    "\n",
    "logPattern = re.compile(\"^log\", flags=re.IGNORECASE)\n",
    "patternPseudoCount = re.compile( \"\\+\\s*[0-9]+(\\.[0-9]+)?\\s*\\)\")\n",
    "patternNumber = re.compile(\"[0-9]+(\\.[0-9]+)?\")\n",
    "logData = unit and re.search(logPattern, unit)\n",
    "pseudocount = (unit and re.search(patternPseudoCount, unit) and float(re.search(patternNumber, re.search(patternPseudoCount, unit).group(0)).group(0))) or 0\n",
    "\n",
    "assert(len(samples) == len(values))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fa471b65",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter code_exec\n",
    "case_group = {{case_group.value}}\n",
    "control_group = {{control_group.value}}\n",
    "\n",
    "gs_library = \"{{gs_library.value}}\"\n",
    "\n",
    "interactive_plot = {{interactive_plot.value}}\n",
    "filter_genes = {{filter_genes.value}}\n",
    "low_expression_threshold = {{low_expression_threshold.value}}\n",
    "nr_genes = {{nr_genes.value}}\n",
    "logCPM_normalization = {{logCPM_normalization}}\n",
    "log_normalization = {{log_normalization}}\n",
    "z_normalization = {{z_normalization}}\n",
    "q_normalization = {{q_normalization}}\n",
    "save_intermediate = {{save_intermediate}}\n",
    "\n",
    "diff_gex_method = \"{{diff_gex_method.value}}\"\n",
    "diff_gex_plot_method = \"{{diff_gex_plot_method.value}}\"\n",
    "pvalue_threshold = {{pvalue_threshold.value}}\n",
    "logfc_threshold = {{logfc_threshold.value}}\n",
    "gene_topk = {{gene_topk.value}}\n",
    "\n",
    "permutation = {{permutation.value}}\n",
    "min_genes = {{min_genes.value}}\n",
    "max_genes = {{max_genes.value}}\n",
    "anchors = {{anchors.value}}\n",
    "symmetric = {{symmetric.value}}\n",
    "n_table = {{n_table.value}}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8fb360da",
   "metadata": {},
   "outputs": [],
   "source": [
    "warnings.filterwarnings('ignore')\n",
    "random.seed(0)\n",
    "pandas2ri.activate()\n",
    "if interactive_plot == True:\n",
    "    plot_type = 'interactive'\n",
    "else:\n",
    "    plot_type = 'static'\n",
    "chart_studio.tools.set_credentials_file(username='mjjeon', api_key='v0rpMa6lhST28Sq7XqtM')\n",
    "results = {}\n",
    "table_counter = 1\n",
    "figure_counter = 1"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "534b2560",
   "metadata": {},
   "source": [
    "# Build Subgroup 1 vs. Subgroup 2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "29a43ff7",
   "metadata": {},
   "outputs": [],
   "source": [
    "meta_df = build_meta_df(samples, values, codes)\n",
    "meta_id_column_name, meta_class_column_name = meta_df.columns\n",
    "meta_subgroup_column_name = \"subgroup\"\n",
    "\n",
    "# build case vs. control subgroups\n",
    "error = check_subgroups (control_group, case_group)\n",
    "if error:\n",
    "    display(Markdown(\"<span style=\\\"background-color:#f2dede\\\">Error! {}.</span>\".format(error)))\n",
    "    sleep(10)\n",
    "    raise SystemExit(1)\n",
    "\n",
    "case_subgroup_name = '+'.join(case_group).replace(\"/\",\"_\")\n",
    "case_subgroup_name = re.sub(r'\\W+', '_', case_subgroup_name) # all nonalphanum change to _\n",
    "control_subgroup_name = '+'.join(control_group)\n",
    "control_subgroup_name = re.sub(r'\\W+', '_', control_subgroup_name) # all nonalphanum change to _\n",
    "meta_df[meta_subgroup_column_name] = np.where(np.isin(meta_df[meta_class_column_name], case_group), case_subgroup_name, np.where(np.isin(meta_df[meta_class_column_name], control_group), control_subgroup_name,''))\n",
    "subgroups = [control_subgroup_name, case_subgroup_name] # CONTROL first, CASE second\n",
    "meta_df = meta_df[meta_df[meta_subgroup_column_name]!= '']\n",
    "\n",
    "table_counter = display_object(table_counter, \"Sample size for each subgroup. The table displays the number of samples in each subgroup.\", meta_df.groupby(meta_subgroup_column_name)[meta_subgroup_column_name].count().to_frame(), istable=True)\n",
    "\n",
    "n_case = len(meta_df[meta_df[meta_subgroup_column_name] == case_subgroup_name])\n",
    "n_control  = len(meta_df[meta_df[meta_subgroup_column_name] == control_subgroup_name])\n",
    "if n_case < 2 or n_control < 2:\n",
    "    display(Markdown(\"<span style=\\\"background-color:#f2dede\\\">Error! Each subgroup needs to have at least 2 samples with gene expression data. \\\"{}\\\" group has {} samples. \\\"{}\\\" group has {} samples.</span>\".format(case_subgroup_name, n_case, control_subgroup_name, n_control)))\n",
    "    print (\"Error!\")\n",
    "    sleep(10)\n",
    "    raise SystemExit(1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7298c11c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# set memory, runtime and sample limit for the analysis\n",
    "website = False\n",
    "def sample_exceeded(sample_size, max_sample):\n",
    "    display(Markdown(\"<span style=\\\"background-color:#f2dede\\\">There are {} samples in the analysis, which has exceeded the {} sample limit on this website. You can use Xena to filter down the samples first.</span>\".format(sample_size, max_sample)))\n",
    "    print (\"Error!\")\n",
    "    sleep(10)\n",
    "    raise SystemExit(1)\n",
    "\n",
    "def time_exceeded(signo, frame): \n",
    "    display(Markdown(\"**The analysis is terminated because it has exceeded the maximum time limit.**\"))\n",
    "    print (\"Error!\")\n",
    "    sleep(10)\n",
    "    raise SystemExit(1)\n",
    "    \n",
    "def set_resource_limit(max_time, max_mem):\n",
    "    soft, hard = resource.getrlimit(resource.RLIMIT_CPU)\n",
    "    resource.setrlimit(resource.RLIMIT_CPU, (max_time, hard))\n",
    "    soft, hard = resource.getrlimit(resource.RLIMIT_AS)\n",
    "    resource.setrlimit(resource.RLIMIT_AS, (max_mem, hard))\n",
    "    signal.signal(signal.SIGXCPU, time_exceeded) \n",
    "\n",
    "if os.path.exists(\"../../analysis.xenahubs.net\"):\n",
    "    website = True\n",
    "\n",
    "if website:\n",
    "    max_time = 60 * 20 # max run time is 20 min\n",
    "    max_mem = 19.9 * 1000 * 1000 * 1000 # max memmory allocated 16GB\n",
    "    max_sample = 2000 # max sample size 2000 \n",
    "    set_resource_limit(max_time, max_mem)\n",
    "\n",
    "    # total number of samples in subgroups\n",
    "    if len(meta_df) > max_sample:\n",
    "        sample_exceeded(len(meta_df), max_sample)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6ac0cd4d",
   "metadata": {},
   "source": [
    "# Load Dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4387f0c9",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Downloading expr data file\n",
    "try:\n",
    "    expr_data_filename = find_remote_file(expr_data_filename)\n",
    "except:\n",
    "    print(f\"Error! Gene expression file not found at '{expr_data_filename}'\")\n",
    "    sleep(10)\n",
    "    raise SystemExit(1)\n",
    "\n",
    "print(\"Downloading expression dataset ... Large dataset will take longer.\")\n",
    "buffer = smart_open.smart_open(expr_data_filename,'r')\n",
    "expr_df = parse_xena_expr(buffer, meta_df[meta_id_column_name].to_list())\n",
    "print(\"Download of expression dataset complete\")\n",
    "\n",
    "# display expr table\n",
    "table_counter = display_object(table_counter, \"Gene expression data. The table displays the input gene expression dataset. Rows represent genes, columns represent samples, and values show the gene expression level.\", expr_df, istable = True)\n",
    "\n",
    "# display meta data table\n",
    "table_counter = display_object(table_counter, \"Metadata. The table displays the metadata associated with the samples in the gene expression dataset. Rows represent samples, columns represent metadata categories and subgroups.\", meta_df, istable = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7d25cdf6",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Match samples between the metadata (subgroup column) and the expression datasets\n",
    "try:\n",
    "    check_df(meta_df, meta_id_column_name)\n",
    "except:\n",
    "    print(f\"Error! Column '{meta_id_column_name}' is not in metadata\")\n",
    "try:\n",
    "    check_df(meta_df, meta_class_column_name)\n",
    "except:\n",
    "    print(f\"Error! Column '{meta_class_column_name}' is not in metadata\")\n",
    "\n",
    "meta_df = meta_df[meta_df[meta_id_column_name].isin(expr_df.columns)] #filter metadata\n",
    "expr_df = expr_df.groupby(expr_df.columns, axis=1).mean() # average across duplicated samples, remove sample duplications\n",
    "expr_df = expr_df.loc[:,meta_df[meta_id_column_name]] #reorder\n",
    "expr_df = expr_df.groupby(expr_df.index).sum() # add all values from the same gene, remove gene duplications\n",
    "\n",
    "try:\n",
    "    assert(meta_df.shape[0]==expr_df.shape[1])\n",
    "except:\n",
    "    print(\"Unexpected! Sample mismatch detected\")\n",
    "    sleep(10)\n",
    "    raise SystemExit(1)\n",
    "\n",
    "n_case = len(meta_df[meta_df[meta_subgroup_column_name] == case_subgroup_name])\n",
    "n_control  = len(meta_df[meta_df[meta_subgroup_column_name] == control_subgroup_name])\n",
    "if n_case < 2 or n_control < 2:\n",
    "    display(Markdown(\"<span style=\\\"background-color:#f2dede\\\">Error! Each subgroup needs to have at least 2 samples with gene expression data. \\\"{}\\\" group has {} samples with gene expression data. \\\"{}\\\" group has {} samples with gene expression data.</span>\".format(case_subgroup_name, n_case, control_subgroup_name, n_control)))\n",
    "    print (\"Error!\")\n",
    "    sleep(10)\n",
    "    raise SystemExit(1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "60d51018",
   "metadata": {},
   "outputs": [],
   "source": [
    "# convert non-HUGO probe_level expr_df to hugo-gene-level expr_df\n",
    "to_hugo = False\n",
    "if probemap_filename:\n",
    "    try:\n",
    "        probemap_filename = find_remote_file(probemap_filename)\n",
    "        probemap_buffer = smart_open.smart_open(probemap_filename,'r')\n",
    "        probemap_df = pd.read_csv(probemap_buffer, index_col = 0, sep = \"\\t\", compression = \"infer\", header = None, comment = \"#\")\n",
    "        if check_probe_level(probemap_df) != \"HUGO\":\n",
    "            # convert expr_df to hugo_gene_level expr_df\n",
    "            expr_df = convert_to_hugo(expr_df, probemap_df)\n",
    "            to_hugo = True\n",
    "    except:\n",
    "        print(f\"Error! Probemap file not found at '{probemap_filename}'\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7666e171",
   "metadata": {},
   "outputs": [],
   "source": [
    "dataset = dict()\n",
    "current_dataset = 'rawdata'\n",
    "dataset[current_dataset] = expr_df\n",
    "dataset['dataset_metadata'] = meta_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f029b9a4",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter markdown\n",
    "{% if filter_genes.value == True %}\n",
    "Filter out low expressed genes \n",
    "{% endif %}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8563aa43",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter code_exec\n",
    "{% if filter_genes.value == True %}\n",
    "## Filter out non-expressed genes\n",
    "expr_df = expr_df.loc[expr_df.sum(axis = 1) > 0, :]\n",
    "\n",
    "## Filter out lowly expressed genes\n",
    "mask_low_vals = (expr_df > low_expression_threshold).sum(axis = 1) > 2\n",
    "expr_df = expr_df.loc[mask_low_vals, :]\n",
    "current_dataset += '+filter_genes'\n",
    "dataset[current_dataset] = expr_df\n",
    "{% endif %}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6ccfd0b0",
   "metadata": {},
   "outputs": [],
   "source": [
    "# display expression data table - hugo converted\n",
    "if (to_hugo):\n",
    "    table_counter = display_object(table_counter, \"Gene expression data. The table displays the quantified gene expression dataset after converting input data to using hugo gene names. Rows represent genes, columns represent samples, and values show the gene expression level.\", dataset[current_dataset], istable = True)\n",
    "    # display(create_download_link(dataset[current_dataset], filename=\"gene_level_expr.csv\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2dcdb42d",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter markdown\n",
    "{% if logCPM_normalization.value == True or log_normalization.value == True or z_normalization.value == True or q_normalization.value == True %}\n",
    "# Normalize gene expression data\n",
    "Normalization methods ({% if logCPM_normalization.value %}count per million (CPM), {% endif %} {% if log_normalization.value %} log transformation, {% endif %} {% if z_normalization.value %} Z normalization, {% endif %} {% if q_normalization.value %}quantile normalization {% endif %}) will be applied to convert raw read counts into informative measures of gene expression and remove factors that affect the analysis.\n",
    "{% endif %}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c11d6694",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter code_exec\n",
    "dataset, normalization = normalize(dataset, current_dataset, {{logCPM_normalization}}, {{log_normalization}}, {{z_normalization}}, {{q_normalization}})\n",
    "\n",
    "{% if logCPM_normalization.value == True or log_normalization.value == True or z_normalization.value == True or q_normalization.value == True %}\n",
    "table_counter = display_object(table_counter, \"Normalized data. The table displays the expression values after normalization.\", dataset[normalization], istable = True)\n",
    "{% if save_intermediate.value == True %}\n",
    "display(create_download_link(dataset[normalization], filename = \"normalized_data.csv\"))\n",
    "{% endif %}\n",
    "{% endif %}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3d6f0195",
   "metadata": {},
   "source": [
    "# QC Plot: Total gene expression analysis"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "84e06746",
   "metadata": {},
   "source": [
    "Total gene expression analysis calculates and displays the total gene expression for each sample in the gene expression dataset, facilitating assessment of the overall quality of the data."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e70b7491",
   "metadata": {},
   "outputs": [],
   "source": [
    "meta_df['sum'] = expr_df.sum().tolist()\n",
    "fig = px.bar(meta_df.sort_values(by = 'sum'), x = meta_id_column_name, y = 'sum')\n",
    "\n",
    "if plot_type == 'interactive':\n",
    "        plotly.offline.iplot(fig)\n",
    "else:\n",
    "    py.image.ishow(fig)\n",
    "figure_counter = display_object(figure_counter, \"The total gene expression for each sample. The figure contains an interactive bar chart which displays the total gene expression (sum) mapped to each sample in the dataset. Additional information for each sample is available by hovering over the bars.\", istable=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "867e162b",
   "metadata": {},
   "source": [
    "# Visualize Samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "680dddfb",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter markdown\n",
    "{% if visualization_method.value == \"PCA\" %}\n",
    "Principal Component Analysis (PCA) (Clark et al. 2011) is a statistical technique used to identify global patterns in high-dimensional datasets. It is commonly used to explore the similarity of biological samples in gene expression datasets. To achieve this, gene expression values are transformed into Principal Components (PCs), a set of linearly uncorrelated features which represent the most relevant sources of variance in the data, and subsequently visualized using a scatter plot.\n",
    "{% endif %}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f0dfaab7",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter code_exec\n",
    "{% if visualization_method.value == \"PCA\" %}\n",
    "method = \"PCA\"\n",
    "{% endif %}\n",
    "{% if visualization_method.value == \"t-SNE\" %}\n",
    "method = \"t-SNE\"\n",
    "{% endif %}\n",
    "\n",
    "# Run analysis\n",
    "results[method] = run_dimension_reduction(dataset = dataset, meta_id_column_name = meta_id_column_name, method = method,\\\n",
    "                         nr_genes = nr_genes, normalization = normalization, plot_type = plot_type)\n",
    "# Display results\n",
    "figure_counter = plot_samples(results[method], meta_id_column_name = meta_id_column_name, meta_class_column_name = meta_class_column_name, counter = figure_counter)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d1f989e4",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter markdown\n",
    "# {{diff_gex_method}} Differential Gene Expression"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2db7bd57",
   "metadata": {},
   "source": [
    "Gene expression signatures are alterations in the patterns of gene expression that occur as a result of cellular perturbations such as drug treatments, gene knock-downs or diseases. They can be quantified using differential gene expression (DGE) methods (Ritchie et al. 2015, Clark et al. 2014), which compare gene expression between two groups of samples to identify genes whose expression is significantly altered in the perturbation. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a03b3967",
   "metadata": {},
   "outputs": [],
   "source": [
    "signatures = get_signatures(subgroups, dataset, normalization, diff_gex_method, meta_subgroup_column_name, meta_id_column_name, filter_genes, logData, pseudocount)\n",
    "\n",
    "for label, signature in signatures.items():\n",
    "    table_counter = display_object(table_counter, \"Differentially expressed genes between {} using {}. The figure displays a browsable table containing the gene expression signature generated from a differential gene expression analysis. Every row of the table represents a gene; the columns display the estimated measures of differential expression.\".format(label, diff_gex_method), signature, istable = True)\n",
    "    display(create_download_link(signature, filename = \"DEG_results_{}.csv\".format(label)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4da97307",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter code_exec\n",
    "{% if diff_gex_method.value == \"limma\" or diff_gex_method.value == \"limma_voom\" or diff_gex_method.value == \"edgeR\" or diff_gex_method.value == \"DESeq2\"%}\n",
    "    \n",
    "{% if diff_gex_plot_method.value == \"volcano\" %}\n",
    "results['volcano_plot'] = {}\n",
    "# Loop through signatures\n",
    "for label, signature in signatures.items():\n",
    "    results['volcano_plot'][label] = run_volcano(signature, label, dataset, pvalue_threshold, logfc_threshold, plot_type)\n",
    "    plot_volcano(results['volcano_plot'][label])\n",
    "    figure_counter = display_object(figure_counter, \"Volcano plot for {}. The figure contains an interactive scatter plot which displays the log2-fold changes and statistical significance of each gene calculated by performing a differential gene expression analysis. Genes with logFC > {} and p-value < {} in red and genes with logFC < -{} and p-value < {} in blue. Additional information for each gene is available by hovering over it.\".format(label, logfc_threshold, pvalue_threshold, logfc_threshold, pvalue_threshold), istable=False)\n",
    "\n",
    "{% elif diff_gex_plot_method.value == \"MA_plot\" %}\n",
    "# Initialize results\n",
    "results['ma_plot'] = {}\n",
    "\n",
    "# Loop through signatures\n",
    "for label, signature in signatures.items():\n",
    "    # Run analysis\n",
    "    results['ma_plot'][label] = run_maplot(signature = signature, signature_label = label, pvalue_threshold = pvalue_threshold, logfc_threshold = logfc_threshold, plot_type = plot_type)\n",
    "    # Display results\n",
    "    plot_maplot(results['ma_plot'][label])\n",
    "    figure_counter = display_object(figure_counter, \"MA plot for {}. The figure contains an interactive scatter plot which displays the average expression and statistical significance of each gene calculated by performing differential gene expression analysis. Genes with logFC > {} and p-value < {} in red and genes with logFC < -{} and p-value < {} in blue. Additional information for each gene is available by hovering over it.\".format(label, logfc_threshold, pvalue_threshold, logfc_threshold, pvalue_threshold), istable=False)\n",
    "\n",
    "{% endif %}\n",
    "{% endif %}"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "77d112ac",
   "metadata": {},
   "source": [
    "# blitzGSEA Full Table"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ab52d028",
   "metadata": {},
   "source": [
    "[blitzGSEA](https://pubmed.ncbi.nlm.nih.gov/35143610/) is an algorithm developed by a group of researchers at Ma'ayan Labs. It has been evaluated to be significantly more performant than the [traditional GSEA](https://www.pnas.org/doi/10.1073/pnas.0506580102) implementation while producing similar results that are biologically consistent. The gene set libraries available to run blitzGSEA with through Xena is retrieved from the [Molecular Signatures Database (MSigDB)](http://www.gsea-msigdb.org/gsea/msigdb/collections.jsp). "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cab80d30",
   "metadata": {},
   "outputs": [],
   "source": [
    "HTML(\"<style>.dataframe td {white-space: nowrap;}</style>\")\n",
    "signature_df = reformat_signature(signatures, diff_gex_method)\n",
    "signature_df.to_csv('signature.tsv', sep='\\t', index=False)\n",
    "\n",
    "n_file = open('n_value.tsv', 'w')\n",
    "n_file.write(str(n_table))\n",
    "n_file.close()\n",
    "\n",
    "display(Markdown(\"### Gene Set Library: {}\".format(gs_library)))\n",
    "file = '/app/data/gsLibrary/' + gs_library\n",
    "library = reformat_gmt(file)\n",
    "full_table = run_blitzGSEA(signature_df, library)\n",
    "full_table.reset_index(inplace=True) \n",
    "    \n",
    "with pd.option_context('display.max_colwidth', 30):\n",
    "    display(full_table.head(n_table))\n",
    "table_counter = display_object(table_counter, \"The table displays the enrichment score (es), normalized enrichment score (nes), p-value (pval), sidak correction (sidak), false discovery rate (fdr), gene set size (geneset_size), and leading genes (leading_edge) for each gene set. Each row is a gene set.\", istable=True)\n",
    "display(create_download_link(full_table, title=\"Download CSV file: {}\", filename=\"blitzGSEA_full_table_{}.csv\".format(gs_library)))\n",
    "\n",
    "full_table.to_csv('full_table.tsv', sep='\\t', index=None)\n",
    "library = pd.read_csv(file)\n",
    "library.to_csv('library.tsv', sep='\\t', index=None)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "917ea1be",
   "metadata": {},
   "source": [
    "# blitzGSEA Top Table, Running Sum Plot, and Detailed Output"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "4e8a580c",
   "metadata": {},
   "source": [
    "Click below to access an interactive top table and view running sum plots and detailed tables of selected gene sets. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6dc84761",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter markdown\n",
    "**[View Gene Set Data Dashboard](../dashboard/{{_session}})**"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "blitzGSEA_appyter",
   "language": "python",
   "name": "blitzgsea_appyter"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
